File: playground/Validate Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */


/*
    這題的玄機在於 , 除了child對節點本身要滿足的大小關係以外 , 他還必須要大於其上源的節點 
    藉著example的圖片可以去觀察 , 我們會在往左右探索的同時去限縮一個區間 ,所有節點必須在區間之內
    用traverse的方式走訪 , maintain這個有效區間
*/
var isValidBST = function(root) {
    
    let valid = true ; 
    // l , r mean the effect value range of the node must be obey 
    const validNode = function(node , l , r){

        if (node == null || !valid){return  ;}

        else if (node.val <= l || node.val >= r){
            valid=false ; 
            return 
        }

        // 往left child走得時候 , 更新right bound 
        // 往right child走的時候 , 更新left bound 
        validNode( node.left  , l  ,  Math.min(r , node.val ));
        validNode(node.right ,  Math.max(l , node.val) , r ) ; 

    }

    validNode(root , Number.MIN_SAFE_INTEGER , Number.MAX_SAFE_INTEGER) ; 

    return valid ; 
};

File: RUSH/Validate Binary Search Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from typing import Optional


# traverse整個tree,並在過程中檢查是否有節點違反BST性質 ,
# 為了做到檢查,在traverse的過程中要maintain左右bound 
# 當我們往左節點找 , 左節點必須小於當前節點 , 因此rightBound要進行更新 
class Solution:

    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        sol = True 


        def valid(node , leftBound , rightBound) : 
            nonlocal sol 
            if node is None : return None 
            if not sol : return 
            
            if leftBound < node.val < rightBound : 
                pass 
            else : 
                sol = False 
                
            # 往左節點走,更新rightBound
            valid(node.left  , leftBound , node.val)
            valid(node.right , node.val , rightBound) 

        valid(root , float('-inf') ,float('inf'))
        
        return sol 
                
            